File: src/1-500/33/RotateSearch.py
from typing import List


class Solution:
    def search(self, nums: List[int], target: int) -> int:
        start, end = 0, len(nums) - 1
        while start <= end:
            mid = (start + end) // 2
            if nums[mid] == target:
                return mid

            if nums[mid] >= nums[start]:
                if nums[mid] > target and nums[start] <= target:
                    end = mid - 1
                else:
                    start = mid + 1
            if nums[mid] < nums[end]:
                if nums[mid] < target and nums[end] >= target:
                    start = mid + 1
                else:
                    end = mid - 1
        return -1

solution = Solution()
print(solution.search([4,5,6,7,0,1,2], 0))
# print(solution.search([4, 5, 6, 7, 0, 1, 2], 3))

File: src/1-500/033/RotateSearch.java
public class Solution {
    public int search(int[] nums, int target) {
        if (nums == null || nums.length == 0) {
            return -1;
        }
        
        int lo = 0; 
        int hi = nums.length - 1;
        
        while(lo + 1 < hi){
            int mid = lo + (hi - lo)/2;
            
            if(nums[mid] == target) return mid;
            
            if(nums[lo] < nums[mid]){
                if(nums[lo] <= target && target < nums[mid]){
                    hi = mid;
                }
                else{
                    lo = mid;
                }
            }
            
            else {
                 if (nums[mid] < target && target <= nums[hi] ) {
                    lo = mid;
                } else {
                    hi = mid;
                }
            }
        }
        
        if (target == nums[lo]) {
            return lo;
        }
         
        if (target == nums[hi]) {
            return hi;
        }
         
        return -1;
    }
}
